_posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd

```diff
@@ -0,0 +1,705 @@
+    rnd_scale <- runif(1, 1 - scale_param, 1 + scale_param)
+    img <- transform_resize(img, size = rnd_scale * img_size)
+    mask <- transform_resize(mask, size = rnd_scale * img_size)
+    diff <- dim(img)[2] - img_size
+    if (diff > 0) {
+      top <- ceiling(diff / 2)
+      left <- ceiling(diff / 2)
+      img <- transform_crop(img, top, left, img_size, img_size)
+      mask <- transform_crop(mask, top, left, img_size, img_size)
+    } else {
+      img <- transform_pad(img,
+                           padding = -c(
+                             ceiling(diff / 2),
+                             floor(diff / 2),
+                             ceiling(diff / 2),
+                             floor(diff / 2)
+                           ))
+      mask <- transform_pad(mask, padding = -c(
+        ceiling(diff / 2),
+        floor(diff /
+                2),
+        ceiling(diff /
+                  2),
+        floor(diff /
+                2)
+      ))
+    }
+    list(img, mask)
+  },
+  
+  rotate = function(img, mask, rot_param) {
+    rnd_rot <- runif(1, 1 - rot_param, 1 + rot_param)
+    img <- transform_rotate(img, angle = rnd_rot)
+    mask <- transform_rotate(mask, angle = rnd_rot)
+    
+    list(img, mask)
+  },
+  
+  flip = function(img, mask, flip_param) {
+    rnd_flip <- runif(1)
+    if (rnd_flip > flip_param) {
+      img <- transform_hflip(img)
+      mask <- transform_hflip(mask)
+    }
+    
+    list(img, mask)
+  }
+)
+
+```
+
+After instantiation, we see we have 2977 training pairs and 952 validation pairs, respectively:
+
+```{r}
+train_ds <- brainseg_dataset(
+  train_dir,
+  augmentation_params = c(0.05, 15, 0.5),
+  random_sampling = TRUE
+)
+
+length(train_ds)
+# 2977
+
+valid_ds <- brainseg_dataset(
+  valid_dir,
+  augmentation_params = NULL,
+  random_sampling = FALSE
+)
+
+length(valid_ds)
+# 952
+```
+
+As a correctness check, let's plot an image and associated mask:
+
+```{r}
+par(mfrow = c(1, 2), mar = c(0, 1, 0, 1))
+
+img_and_mask <- valid_ds[27]
+img <- img_and_mask[[1]]
+mask <- img_and_mask[[2]]
+
+img$permute(c(2, 3, 1)) %>% as.array() %>% as.raster() %>% plot()
+mask$squeeze() %>% as.array() %>% as.raster() %>% plot()
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/scan_and_mask.png")
+```
+
+With `torch`, it is straightforward to inspect what happens when you change augmentation-related parameters. We just pick a pair from the validation set, which has not had any augmentation applied as yet, and call `valid_ds$<augmentation_func()>` directly. Just for fun, let's use more "extreme" parameters here than we do in actual training. (Actual training uses the settings from Mateusz' GitHub repository, which we assume have been carefully chosen for optimal performance.[^1])
+
+[^1]: Yes, we did a few experiments, confirming that more augmentation isn't better ... what did I say about inevitably ending up doing optimization on the validation set ...
```

Comment:
I think you want a question mark at the end of the question rather than a second ellipsis.

Reply:
ok!